diff --git a/go.mod b/go.mod
index 19136dcf..4cc4f294 100644
--- a/go.mod
+++ b/go.mod
@@ -2,6 +2,8 @@ module tailscale.com
 
 go 1.20
 
+replace github.com/tailscale/wireguard-go => ../wireguard-go
+
 require (
 	filippo.io/mkcert v1.4.3
 	github.com/Microsoft/go-winio v0.6.0
diff --git a/net/netcheck/netcheck.go b/net/netcheck/netcheck.go
index d53eb7ee..fff25dae 100644
--- a/net/netcheck/netcheck.go
+++ b/net/netcheck/netcheck.go
@@ -1106,7 +1106,7 @@ func (c *Client) checkCaptivePortal(ctx context.Context, dm *tailcfg.DERPMap, pr
 		(preferredDERP != 0 && len(dm.Regions[preferredDERP].Nodes) == 0) {
 		rids := make([]int, 0, len(dm.Regions))
 		for id, reg := range dm.Regions {
-			if reg == nil || reg.Avoid || len(reg.Nodes) == 0 {
+			if reg == nil || reg.Avoid || len(reg.Nodes) == 0 || reg.Nodes[0].DERPPort != 0 {
 				continue
 			}
 			rids = append(rids, id)
@@ -1125,8 +1125,11 @@ func (c *Client) checkCaptivePortal(ctx context.Context, dm *tailcfg.DERPMap, pr
 		// TODO(bradfitz,andrew-d): how to actually handle this nicely?
 		return false, nil
 	}
-
-	req, err := http.NewRequestWithContext(ctx, "GET", "http://"+node.HostName+"/generate_204", nil)
+	httpport := 80
+	if node.DERPPort != 0 {
+		httpport = node.DERPPort
+	}
+	req, err := http.NewRequestWithContext(ctx, "GET", "http://"+node.HostName+":"+fmt.Sprint(httpport)+"/generate_204", nil)
 	if err != nil {
 		return false, err
 	}
@@ -1177,7 +1180,11 @@ func (c *Client) runHTTPOnlyChecks(ctx context.Context, last *Report, rs *report
 		go func() {
 			defer wg.Done()
 			node := rg.Nodes[0]
-			req, _ := http.NewRequestWithContext(ctx, "HEAD", "https://"+node.HostName+"/derp/probe", nil)
+			tlsport := 443
+			if node.DERPPort != 0 {
+				tlsport = node.DERPPort
+			}
+			req, _ := http.NewRequestWithContext(ctx, "HEAD", "https://"+node.HostName+":"+fmt.Sprint(tlsport)+"/derp/probe", nil)
 			// One warm-up one to get HTTP connection set
 			// up and get a connection from the browser's
 			// pool.
@@ -1248,8 +1255,11 @@ func (c *Client) measureHTTPSLatency(ctx context.Context, reg *tailcfg.DERPRegio
 		},
 	}
 	hc := &http.Client{Transport: tr}
-
-	req, err := http.NewRequestWithContext(ctx, "GET", "https://"+node.HostName+"/derp/latency-check", nil)
+	tlsport := 443
+	if node.DERPPort != 0 {
+		tlsport = node.DERPPort
+	}
+	req, err := http.NewRequestWithContext(ctx, "GET", "https://"+node.HostName+":"+fmt.Sprint(tlsport)+"/derp/latency-check", nil)
 	if err != nil {
 		return 0, ip, err
 	}
@@ -1448,6 +1458,11 @@ func (c *Client) addReportHistoryAndSetPreferredDERP(r *Report) {
 	var bestAny time.Duration
 	var oldRegionCurLatency time.Duration
 	for regionID, d := range r.RegionLatency {
+		if regionID < 0 {
+			// STUNOnly is not readable here
+			// so use negative region IDs to mark stun-only servers
+			continue
+		}
 		if regionID == prevDERP {
 			oldRegionCurLatency = d
 		}
diff --git a/net/tsaddr/tsaddr.go b/net/tsaddr/tsaddr.go
index 34259b69..788df2bf 100644
--- a/net/tsaddr/tsaddr.go
+++ b/net/tsaddr/tsaddr.go
@@ -24,6 +24,18 @@ func ChromeOSVMRange() netip.Prefix {
 
 var chromeOSRange oncePrefix
 
+func CustomNetRange() netip.Prefix {
+	customNetRange.Do(func() { mustPrefix(&customNetRange.v, "172.22.98.0/24") })
+	return customNetRange.v
+}
+func CustomNet6Range() netip.Prefix {
+	customNet6Range.Do(func() { mustPrefix(&customNet6Range.v, "fd00:3d22:4e98:cafe::/64") })
+	return customNet6Range.v
+}
+
+var customNetRange oncePrefix
+var customNet6Range oncePrefix
+
 // CGNATRange returns the Carrier Grade NAT address range that
 // is the superset range that Tailscale assigns out of.
 // See https://tailscale.com/s/cgnat
@@ -69,9 +81,9 @@ func TailscaleServiceIPv6() netip.Addr {
 // Tailscale assigns from.
 func IsTailscaleIP(ip netip.Addr) bool {
 	if ip.Is4() {
-		return CGNATRange().Contains(ip) && !ChromeOSVMRange().Contains(ip)
+		return CustomNetRange().Contains(ip) || (CGNATRange().Contains(ip) && !ChromeOSVMRange().Contains(ip))
 	}
-	return TailscaleULARange().Contains(ip)
+	return CustomNet6Range().Contains(ip) || TailscaleULARange().Contains(ip)
 }
 
 // TailscaleULARange returns the IPv6 Unique Local Address range that
diff --git a/wgengine/wgcfg/nmcfg/nmcfg.go b/wgengine/wgcfg/nmcfg/nmcfg.go
index d07f7232..e16671aa 100644
--- a/wgengine/wgcfg/nmcfg/nmcfg.go
+++ b/wgengine/wgcfg/nmcfg/nmcfg.go
@@ -102,6 +102,17 @@ func WGCfg(nm *netmap.NetworkMap, logf logger.Logf, flags netmap.WGConfigFlags,
 
 		didExitNodeWarn := false
 		cpeer.V4MasqAddr = peer.SelfNodeV4MasqAddrForThisPeer
+		var localIps []netip.Addr
+		if nm.SelfNode != nil {
+			for _, ep := range nm.SelfNode.Endpoints {
+				addr, err := netip.ParseAddrPort(ep)
+				if err == nil {
+					localIps = append(localIps, addr.Addr())
+				} else {
+					logf("[v1] wgcfg: unable to parse endpoint %q: %v", ep, err)
+				}
+			}
+		}
 		for _, allowedIP := range peer.AllowedIPs {
 			if allowedIP.Bits() == 0 && peer.StableID != exitNode {
 				if didExitNodeWarn {
@@ -127,6 +138,21 @@ func WGCfg(nm *netmap.NetworkMap, logf logger.Logf, flags netmap.WGConfigFlags,
 					}
 					fmt.Fprintf(skippedSubnets, "%v from %q (%v)", allowedIP, nodeDebugName(peer), peer.Key.ShortString())
 					continue
+				} else {
+					ignoreRoute := false
+					for _, localIP := range localIps {
+						if allowedIP.Contains(localIP) {
+							ignoreRoute = true
+							break
+						}
+					}
+					if ignoreRoute {
+						if skippedSubnets.Len() > 0 {
+							skippedSubnets.WriteString(", ")
+						}
+						fmt.Fprintf(skippedSubnets, "%v (localip) from %q (%v)", allowedIP, nodeDebugName(peer), peer.Key.ShortString())
+						continue
+					}
 				}
 			}
 			cpeer.AllowedIPs = append(cpeer.AllowedIPs, allowedIP)
